Create a `mode: time_series` option for indices

nik9000 opened the pull request:
Enabling this setting will put the index into a mode optimized for time
series data, grouping documents who's fields are annotated with
`dimension`. Because time time series data is usually write-once we also
disable your ability to update or delete documents.

We have big plans for things we can do with data organized in this way,
but for now the primary advantage of enabling `time_series_mode` is that
it shrinks the index on disk. We did some local tests with k8s
monitoring data in ECS format. Here are some figures for a single
sharded index without replicas, force merged to a single segment:
```
                    uncompressed JSON: 150GB
                             defaults:  60GB
                     best_compression:  39GB
                     time_series_mode:  33GB
best_compression and time_series_mode:  22GB
```


So the compression you get is similar to best_compression and it can
operate "on top of" best compression to further reduce the index size.

The price is, like we mentioned above, disabling update and delete of
documents. In addition to that, there is a performance cost on ingest
and on mere. Loading the data from above took:
```
                              Index time     Force merge time     Size after force merge
                       master 2.6 hours      0.8 hours            39.3gb
  master sorted by dimensions 2.9 hours      2.2 hours            21.3gb
tsid_gen timeseries_mode=true 2.6 hours      1.7 hours            21.8gb
```

In addition to the space savings this creates an unsearchable `_tsid`
field which functions as a short hand for aggregating on all the
dimension. So you can run this:
```
POST localhost:9200/test/_search?pretty
{
  "size": 0,
  "aggs": {
    "tsid": {
      "terms": { "field": "_tsid" },
      "aggs": {
        "max_n": { "max": { "field": "n" } }
      }
    }
  }
}
```

and the `key` field in the json of the terms bucket will contain all of
the dimensions. Something like this:

```
"aggregations" : {
  "tsid" : {
    "buckets" : [
      {"key":{"dim1":"a","dim2":"a"},"doc_count":6,"max_n":{"value":6}},
      {"key":{"dim1":"a","dim2":"b"},"doc_count":6,"max_n":{"value":6}},
      {"key":{"dim1":"a","dim2":"c"},"doc_count":6,"max_n":{"value":6}},
      {"key":{"dim1":"a","dim2":"d"},"doc_count":6,"max_n":{"value":6}},
      {"key":{"dim1":"b","dim2":"a"},"doc_count":6,"max_n":{"value":6}},
      {"key":{"dim1":"b","dim2":"b"},"doc_count":6,"max_n":{"value":6}},
      {"key":{"dim1":"b","dim2":"c"},"doc_count":6,"max_n":{"value":6}},
      {"key":{"dim1":"b","dim2":"d"},"doc_count":6,"max_n":{"value":6}},
      {"key":{"dim1":"c","dim2":"a"},"doc_count":6,"max_n":{"value":6}},
      {"key":{"dim1":"c","dim2":"b"},"doc_count":6,"max_n":{"value":6}}
    ]
  }
}
```

nik9000 committed:
Create a `time_series_mode` option for indices

Enabling this setting will put the index into a mode optimized for time
series data, grouping documents who's fields are annotated with
`dimension`. Because time time series data is usually write-once we also
disable your ability to update or delete documents.

We have big plans for things we can do with data organized in this way,
but for now the primary advantage of enabling `time_series_mode` is that
it shrinks the index on disk. We did some local tests with k8s
monitoring data in ECS format. Here are some figures for a single
sharded index without replicas, force merged to a single segment:
                    uncompressed JSON: 150GB
                             defaults:  60GB
                     best_compression:  39GB
                     time_series_mode:  33GB
best_compression and time_series_mode:  22GB
((NOCOMMIT: get the size measurements after running the speed tests))

So the compression you get is similar to best_compression and it can
operate "on top of" best compression to further reduce the index size.

The price is, like we mentioned above, disabling update and delete of
documents. In addition to that, there is a performance cost on ingest
and on mere. Loading the data from above took:
                                          <load>      <force_merge>
                       best_compression:  2.6 hours   0.8 hours
best_compression sort on all dimensions:
                       time_series_mode:  (measuring)
  best_compression and time_series_mode:  (measuring)

In addition to the space savings this creates an unsearchable `_tsid`
field which functions as a short hand for aggregating on all the
dimension. So you can run this:
```
POST localhost:9200/test/_search?pretty
{
  "size": 0,
  "aggs": {
    "tsid": {
      "terms": { "field": "_tsid" },
      "aggs": {
        "max_n": { "max": { "field": "n" } }
      }
    }
  }
}
```

and the `key` field in the json of the terms bucket will contain all of
the dimensions. Something like this:

```
"aggregations" : {
  "tsid" : {
    "buckets" : [
      {"key":{"dim1":"a","dim2":"a"},"doc_count":6,"max_n":{"value":6}},
      {"key":{"dim1":"a","dim2":"b"},"doc_count":6,"max_n":{"value":6}},
      {"key":{"dim1":"a","dim2":"c"},"doc_count":6,"max_n":{"value":6}},
      {"key":{"dim1":"a","dim2":"d"},"doc_count":6,"max_n":{"value":6}},
      {"key":{"dim1":"b","dim2":"a"},"doc_count":6,"max_n":{"value":6}},
      {"key":{"dim1":"b","dim2":"b"},"doc_count":6,"max_n":{"value":6}},
      {"key":{"dim1":"b","dim2":"c"},"doc_count":6,"max_n":{"value":6}},
      {"key":{"dim1":"b","dim2":"d"},"doc_count":6,"max_n":{"value":6}},
      {"key":{"dim1":"c","dim2":"a"},"doc_count":6,"max_n":{"value":6}},
      {"key":{"dim1":"c","dim2":"b"},"doc_count":6,"max_n":{"value":6}}
    ]
  }
}
```

nik9000 commented on server/src/main/java/org/elasticsearch/action/index/IndexRequest.java:
```diff
@@ -605,8 +613,56 @@ public void process(Version indexCreatedVersion, @Nullable MappingMetadata mappi
+                    "alias routing incompatible the destination index [" + abstraction.getName() + "] because it is in time series mode"
+                );
+            }
+            routing(routingFromTimeSeries(abstraction, timeSeriesGeneratorLookup));
```
I'd like to put this into its own field and modify the routing behavior to pick it up. But I've not done that in this PR because I think it's fairly mechanical and change and I believe I'd have to disable a lot of BWC tests for it. I'd like to do that in a follow up. So instead I'm hiding the tsid in the _routing field, base 64 encoding it. Ewwwwww. But it works and we can zap it in a follow up.

jpountz replied:
Maybe you already looked into it, but this might require some changes for `_split` to keep working with time series indices if we did that. `_split` looks at the value of the `_routing` field to know how to split shards in a way that still honors the rounting key (see `ShardSplittingQuery`).

nik9000 replied:
OOOH! I hadn't thought of it. You've given me a wonderful thing to experiment with. I imagine I break this now. I'll look!

nik9000 replied:
Shard splitting query is indeed busted here. We'll need revive it, but I'd like to wait for a follow up for that.

nik9000 commented on server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:
```diff
@@ -558,6 +558,7 @@ public String resolveWriteIndexRouting(@Nullable String routing, String aliasOrI
             return routing;
         }
 
+        // TODO this is the same sort of code we have in bulk action already. we should share
```
Over the years we've grown a lot of very similar code over in TransportBulkAction. We should unify these. *somehow*.

nik9000 commented on server/src/main/java/org/elasticsearch/common/settings/IndexScopedSettings.java:
```diff
@@ -48,7 +49,7 @@
     public static final Predicate<String> INDEX_SETTINGS_KEY_PREDICATE = (s) -> s.startsWith(IndexMetadata.INDEX_SETTING_PREFIX);
 
-    public static final Set<Setting<?>> BUILT_IN_INDEX_SETTINGS = Set.of(
+    private static final Set<Setting<?>> ALWAYS_ENABLED_BUILT_IN_INDEX_SETTINGS = Set.of(
```
This makes sure that the index setting is only registered if the feature flag is enabled or we're on a snapshot. I tried to do this in the way that "looks small" in the diff so it's easier for us to reason about....

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/IpFieldMapper.java:
```diff
@@ -443,34 +447,28 @@ protected String contentType() {
 
     @Override
     protected void parseCreateField(DocumentParserContext context) throws IOException {
```
I wanted very much to reuse the parsing code but it was super twisty. I believe the new code doesn't change the behavior but let's us reuse it during tsid extraction.

nik9000 commented on server/src/main/java/org/elasticsearch/index/query/CoordinatorRewriteContext.java:
```diff
@@ -58,6 +58,7 @@ boolean hasTimestampData() {
 
     @Nullable
     public MappedFieldType getFieldType(String fieldName) {
+        // TODO use time series id generation?
```
It'd be complex, but we already have the time series id generation code on the coordinating node. So - if you are querying all of the dimensions - we could generate the time series id that must match here and then use the routing it makes to limit the nodes we land on.

nik9000 commented on server/src/main/java/org/elasticsearch/indices/IndicesService.java:
```diff
@@ -233,15 +234,7 @@
     private final ValuesSourceRegistry valuesSourceRegistry;
     private final TimestampFieldMapperService timestampFieldMapperService;
     private final CheckedBiConsumer<ShardSearchRequest, StreamOutput, IOException> requestCacheKeyDifferentiator;
-
```
I moved this method below the ctor and added the time series id generator service as a cluster state consumer. Having instance methods above the ctor confuses me.

nik9000 commented on server/src/main/java/org/elasticsearch/indices/IndicesService.java:
```diff
@@ -1702,4 +1711,7 @@ public boolean allPendingDanglingIndicesWritten() {
         return timestampFieldMapperService.getTimestampFieldType(index);
     }
 
+    public Function<IndexMetadata, TimeSeriesIdGenerator> getTimeSeriesGeneratorLookup() {
```
I use the `Function` here so its easier to test the bulk action without standing up a bunch of stuff.

nik9000 commented on server/src/main/java/org/elasticsearch/indices/TimeSeriesIdGeneratorService.java:
```diff
@@ -0,0 +1,386 @@
+class TimeSeriesIdGeneratorService extends AbstractLifecycleComponent
+    implements
+        ClusterStateApplier,
+        Function<IndexMetadata, TimeSeriesIdGenerator> {
```
There's a fair bit of tricky visibility stuff going on here. I'm sorry.

nik9000 replied:
It should *feel* similar to `TimestampFieldMapperService` but with different constraints - instead of returning empty if we lose the race we calculate the result in the calling thread.

nik9000 commented on server/src/main/java/org/elasticsearch/indices/TimeSeriesIdGeneratorService.java:
```diff
@@ -0,0 +1,386 @@
+     * mapping's {@link CompressedXContent} but {@link CompressedXContent#equals(Object)}
+     * will try to decompress the mapping if the crc matches but the compressed bytes
+     * don't. That's wasteful for us - probably for everyone. If the crc and compressed
+     * bytes match that's a match.
```
It's wasteful specifically because if we don't match we need to decompress the bytes anyway. Since that's slow we don't want to decompress a second time. I'm fairly sure it'd be safe to modify the `equals` implementation in `CompressedXContent` but I don't want to tempt fate in this PR.

nik9000 commented on server/src/main/java/org/elasticsearch/indices/TimeSeriesIdGeneratorService.java:
```diff
@@ -0,0 +1,386 @@
+
+        @Override
+        public int hashCode() {
+            return mapping.hashCode();
```
`CompressedXContent#hashCode` is fast so its safe to just delegate to it.

nik9000 commented on server/src/main/java/org/elasticsearch/search/DocValueFormat.java:
```diff
@@ -655,4 +657,35 @@ public double parseDouble(String value, boolean roundUp, LongSupplier now) {
+        @Override
+        public Object format(BytesRef value) {
+            try {
+                return TimeSeriesIdGenerator.parse(new BytesArray(value).streamInput());
```
I chose to make `parse` static here so any code can decode a `_tsid` without knowing what mapping it came from. We could go the other way - it'd save a few bytes of encoding. But I don't think its worth it.

nik9000 commented on server/src/test/java/org/elasticsearch/index/IndexSortSettingsTests.java:
```diff
@@ -135,13 +141,8 @@ public void testInvalidMissing() {
     }
 
-    public void testIndexSorting() {
+    public void testIndexSortingNoDocValues() {
```
I just renamed this method. I didn't change what it does. All of the removed lines are just relocated into `buildIndexSort`.

nik9000 commented on server/src/test/java/org/elasticsearch/index/mapper/DocumentMapperTests.java:
```diff
@@ -284,8 +295,59 @@ public void testEmptyDocumentMapper() {
         assertNotNull(documentMapper.sourceMapper());
         assertNotNull(documentMapper.IndexFieldMapper());
-        assertEquals(10, documentMapper.mappers().getMapping().getMetadataMappersMap().size());
-        assertEquals(10, documentMapper.mappers().getMatchingFieldNames("*").size());
```
These assertions failed in my PR but were hard to read. So I made them easier to read while fixing them.

nik9000 commented:
@elasticsearchmachine run elasticsearch-ci/part-2

nik9000 committed:
Better support for feature flag

weizijun commented:
Wow, you are so efficient!

nik9000 commented:
Not so much! We've been talking about this for a while. And I was working
on this for a while too!

On Thu, Jul 22, 2021, 23:25 weizijun ***@***.***> wrote:

> Wow, you are so efficient!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/75638#issuecomment-885373174>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABUXIX5WMJE26EWFCQYO7LTZDOKJANCNFSM5A2M2XQQ>
> .
>

hendrikmuhs commented:
Great stuff!

I have 2 questions:

- What's the reason - if I read correctly - _not_ to sort the index on `@timestamp` 1st and 2nd on `_tsid`?
- I wonder if it would be beneficial to make `dimension` hierarchic, meaning instead of `true/false` you configure an order class and sort is applied in the given order. Data usually has such hierarchies like that, with prefix search this could be uses in optimizations.

nik9000 commented:
>     * What's the reason - if I read correctly - _not_ to sort the index on `@timestamp` 1st and 2nd on `_tsid`?

That's a great question! One who's answer I'll want to stick in the javadocs somewhere. But I'll reply here too. We sort on tsid first because all the things we want to do work better that way. Well, two things I can think of:
1. It makes it very easy to, at search time, iterate on a single time series in timestamp order. You do the traditional heap of iterators - this time the iterator is the reader, the timeseries id, and the timestamp. This makes it *much* easier to implement certain features like getting the rate of change of a counter.
2. _source compression works better when you group a single time series together tightly. A lot better. This has a fair bit to do with the "shape" of the data that folks put in ES for time series data right now. It contains a *lot* of "tags" that are mostly constant. I don't have a real example handy, but its stuff like, in data about a volume mounted on a k8s pod we log the pod's IP and creation date and which physical node is hosting it. Stuff that doesn't uniquely identify the data but it is important to be able to see. When you sort by timeseries id first lz4 can do a great job compressing these duplicates. 


@imotov I sorted by `_tsid`, `@timestamp`, both in *ascending* order. Is that right for you? I believe for the compression the choice of order doesn't matter. But you've thought a lot more about the features we can build on this. Do they need `@timestamp` in ascending order? Or is descending order OK? I get the sense that doing `@timestamp` in descending order is going to help a lot of the traditional ES operations like fetching the last 15 minutes of data. Or last hour.

jpountz commented:
> I wonder if it would be beneficial to make dimension hierarchic, meaning instead of true/false you configure an order class and sort is applied in the given order. Data usually has such hierarchies like that, with prefix search this could be used in optimizations.

I like the simplicity of the binary choice for dimension fields, though I had wondered about this too for the case of one high-cardinality dimension and another low-cardinality dimension since I *think* we'd generally prefer sorting by the low-cardinality dimension first and then by the high-cardinality dimension. Maybe one way to make this work (in a follow-up PR, this one is complex already) would be to allow configuring an index sort on timeseries indices, but only using dimension fields. And then the timeseries mode would add `_tsid` and `@timestamp` as tiebreakers automatically.

nik9000 commented:
> I like the simplicity of the binary choice for dimension fields, though I had wondered about this too for the case of one high-cardinality dimension and another low-cardinality dimension since I _think_ we'd generally prefer sorting by the low-cardinality dimension first and then by the high-cardinality dimension. Maybe one way to make this work (in a follow-up PR, this one is complex already) would be to allow configuring an index sort on timeseries indices, but only using dimension fields. And then the timeseries mode would add `_tsid` and `@timestamp` as tiebreakers automatically.

I like the simplicity too! Being able to configure index sorting with special rules could do the trick. I think its safe to keep this in mind but experiment more with it later.

@weizijun had an similar feeling proposal: allow the users to specify which dimensions are used in routing. If you don't use a dimension in routing then it is *just* used for index sorting. It'd be useful to force similar documents even closer together. Though I don't know how space savings it'd buy. @weizijun, do you have some experimental results?

I'd like to throw this out there: Traditionally Elasticsearch routes documents to a particular index based on the `_id`. You can override that by setting the routing explicitly. We'll hash *that* instead of the id based routing. Whenever folks do that we warn them that they could unbalance the distribution of the documents. They could end up with shard 1 much bigger than shard 0. This makes Elasticsearch work a lot less well. We assume that shards are fairly similarly sized and have fairly similar cost to search.

This PR will cause more unbalanced shards. For sure. But not as many as just going crazy with routing would. Because it scatters different time series to different nodes randomly. We *do* expect for some time series to get more documents than others. But we expect there to be many many time series so we expect that in most cases the more chatty time series will get scattered to the different shards evenly.

Now, looping around to the routing proposal: it feels like *exactly* the kind of thing that'd make it easy to accidentally route the high traffic time series indices to the same shard. That doesn't mean its not a good idea too. But, like giving folks more control over the sort order, its something I want to do later. And *carefully*. Or not at all.

nik9000 committed:
Merge branch 'master' into tsid_gen

nik9000 committed:
build build build

imotov commented:
> @imotov I sorted by `_tsid`, `@timestamp`, both in _ascending_ order.

That's a good question. Since we will be returning data in the ascending order, it might be easier for us to iterate over the data in the ascending order as well, in this case we don't have to flip it later on. 

I don't think ordering of `_tsid` would make any difference as long as all records with the same `_tsid` are grouped together somehow, so I don't see any reason for not making it ascending to keep things simple.

> I wonder if it would be beneficial to make dimension hierarchic

I agree with @jpountz and @nik9000, things are already very complicated as they are, but I also agree with @hendrikmuhs that indeed data have hierarchical structures. I just want to add a caveat that naturally occurring hierarchies do not always match optimizations. 

Let's take for example `node_cpu_seconds_total` from prometheus node exporter. It has 4 dimensions: cpu, instance, job and mode:

```
node_cpu_seconds_total{cpu="0", instance="node_exporter:9100", job="nodeexporter", mode="idle"}
node_cpu_seconds_total{cpu="0", instance="node_exporter:9100", job="nodeexporter", mode="iowait"}
node_cpu_seconds_total{cpu="0", instance="node_exporter:9100", job="nodeexporter", mode="irq"}
node_cpu_seconds_total{cpu="0", instance="node_exporter:9100", job="nodeexporter", mode="nice"}
node_cpu_seconds_total{cpu="0", instance="node_exporter:9100", job="nodeexporter", mode="softirq"}
node_cpu_seconds_total{cpu="0", instance="node_exporter:9100", job="nodeexporter", mode="steal"}
node_cpu_seconds_total{cpu="0", instance="node_exporter:9100", job="nodeexporter", mode="system"}
node_cpu_seconds_total{cpu="0", instance="node_exporter:9100", job="nodeexporter", mode="user"}
```

So a natural hierarchy would be job, instance, cpu and then mode. However, a typical query would be most likely interested in average non idle cpu rate `sum(1-rate(node_cpu_seconds_total{mode="idle"}[1m]))` so, the most likely optimization would be to filter by mode and then process all data inside, which will be adversary scenario for the most logical hierarchy. Or perhaps if we are interested in only one machine the natural hierarchy would indeed help us. 

I think this idea is definitely worth exploring, but I would like to spend more time to work with this type of data before we start working on this sort of optimization. I just don't think we will get it right at the moment.

nik9000 commented:
> I don't see any reason for not making it ascending to keep things simple.

I'll drop that into a comment. I think its a thing we *could* change our minds on in the future if we need to.

nik9000 committed:
Merge branch 'master' into tsid_gen

nik9000 committed:
Support time series index without dimensions

You can't write any documents into the index, but you can create one.
This is useful because you might want to create an index and then use
dynamic mappings and templates to fill in the dimensions later.

nik9000 commented:
I've updated the timings in the PR. I don't have a rally track for this (yet) so I'm taking some fairly rough numbers. Indexing my data takes *about* 2.6 hours with or without timeseries mode. It takes about 10% longer if you sort of all timensions (2.9 hours) instead of generating the time series id and sorting by that. We like generating the timeseries and anyway because we want to aggregate and route on out - but its nice to know we get some time savings by sorting on it.

Similarly, we get time savings when you force merge down to a single segment. The force merge with just best_compression takes .8 hours. In time series mode it takes 110% longer (1.7 hours). This seems to be dominated by merging the sorted segments. If you just use best_compression the segments aren't sorted at all so this isn't surprising. The force_merge time when you sort on every segment is 2.2 hours. That's almost as long as it takes to index the data with best_compression! Still, these show areas for improvement - merging sorted segments is slower than we'd like.

Finally, I got some new size estimates. Remember, this data uncompressed is 150GB. I keep it compressed with zstd and it clocks in at 18gb. gzipped it is 21gb. When you index it with best_compression a single shard is, like we said above, 39.3gb. That's with the search index, doc values, and the _source compressed with zlib. Doing it in time series mode it is 21.8gb, pretty close to gzip even though we still have the search index and doc values! That's lovely. But we can do better. We will do better. In a follow up.

elasticmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

nik9000 committed:
Split

nik9000 commented:
I've converted this back to a draft because I've discovered some "fun" stuff around recoveries. The trick I do with putting the tsid in to the `routing` doesn't seem to survive recoveries and the translog. I'm investigating.

imotov commented on server/src/main/java/org/elasticsearch/index/TimeSeriesIdGenerator.java:
```diff
@@ -0,0 +1,334 @@
+            Collections.sort(dimensionNames);
+            throw new IllegalArgumentException("Document must contain one of the dimensions " + dimensionNames);
+        }
+        Collections.sort(values, Comparator.comparing(Map.Entry::getKey));
```
I think there is a dedicated method for that - `Map.Entry.comparingByKey()`

imotov commented on server/src/main/java/org/elasticsearch/index/TimeSeriesIdGenerator.java:
```diff
@@ -0,0 +1,334 @@
+        root.extract(values, "", parser);
+        if (values.isEmpty()) {
+            List<String> dimensionNames = new ArrayList<>();
+            root.collectDimensionNames("", dimensionNames::add);
```
You can probably move this to constructor, you have to run collectDimensionNames anyway.

nik9000 replied:
I didn't want keep the string sitting around though. In case there are zillions of these things.

imotov reviewed (commented):
Nice! I really like where it is going. Left a few minor comments. 

I noticed that there are 11 TODOs in this PR, are you planning to fulfill some of them in this PR or they all are for the follow up PRs?

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java:
```diff
@@ -26,14 +28,15 @@ public static DocumentMapper createEmpty(MapperService mapperService) {
         this.type = mapping.getRoot().name();
         this.mappingLookup = MappingLookup.fromMapping(mapping);
         this.mappingSource = mapping.toCompressedXContent();
+        timeSeriesIdGenerator = inTimeSeriesMode ? mapping.buildTimeSeriesIdGenerator() : null;
```
I'll play with moving this. @romseygeek suggested moving it to `MappingLookup` I'll try both and see how I feel.

weizijun commented:
> @weizijun had an similar feeling proposal: allow the users to specify which dimensions are used in routing. If you don't use a dimension in routing then it is _just_ used for index sorting. It'd be useful to force similar documents even closer together. Though I don't know how space savings it'd buy. @weizijun, do you have some experimental results?

We use custom routing, the main purpose is to query fewer shards. Our index have a large number of shards, almost 200, so we hope to query by routing. routing may cause unbalanced shards, we solve it by setting index.routing_partition_size

nik9000 commented:
> We use custom routing, the main purpose is to query fewer shards. Our index have a large number of shards, almost 200, so we hope to query by routing. routing may cause unbalanced shards, we solve it by setting index.routing_partition_size

Gotcha. `index.routing_partition_size` is something I just discovered. And disabled when the index is in this mode. It'd be a good follow up to get it working again, I think.

nik9000 commented:
Update on this:

As written this doesn't work with peer or translog discovery. I had a crash coarse in all that yesterday morning with @DaveCTurner. We decided a few things:
1. It's good to use the `_tsid` that we generate in regular indexing. That works already. :+1:
2. We *don't* want to save the `_tsid` in the translog. We fsync that thing and any extra bytes are bad. We will instead regenerate the `_tsid` when we work off of the translog.
3. Confusingly, peer recovery and cross cluster replication work by building a translog entry out the peer's lucene index. In this case we *do* have the `_tsid` available. But because this is literally the same thing we fsync we don't have a slot to put it in. So during peer recovery and we also will regenerate the _tsid on the fly.

nik9000 committed:
Ouch

nik9000 commented:
I've pushed a fairly large change (+1438 -333) to handle recovery and to add more tests. Now we have a bunch of unit tests with indices in time series mode recovery and replying transations and stuff. I've added a few more markers for extra smoke tests against things like security and cross cluster replication. I'll also need to add tests for full cluster and rolling restarts. Those all should work because all of the unit tests around recovery work in tsdb mode.

I haven't made any changes from code review yet. I did relocate the `timeSeriesIdGenerator` field from `DocumentMapper` to `Mapping` as @imotov suggested as part of getting recovery working. I'm not sure whether I like it better there or not. But there it is. That move did force me to write more unit tests of the system in tsdb mode which feel like a good thing.



One thing that's jumped out at me: you can't create a tsdb index with a fully empty mapping. You need the `@timestamp` field declared in the mapping to or else we can't stand up the index sorting.

nik9000 committed:
Merge branch 'master' into tsid_gen

nik9000 committed:
Tests

nik9000 commented:
I've finished writing a bunch of smoke tests for things like CCR, security, full cluster restart, and rolling restart. I'll pick up the reviews again on Monday and strike draft from this once I've finished with the reviews.

nik9000 committed:
Merge branch 'master' into tsid_gen

weizijun commented:
> One thing that's jumped out at me: you can't create a tsdb index with a fully empty mapping. You need the `@timestamp` field declared in the mapping to or else we can't stand up the index sorting.

Can you consider using time_series_mode in datastream?

nik9000 commented:
It's certainly worth a test to make sure you can make a data stream in the
series mode.

On Sun, Aug 1, 2021, 21:55 weizijun ***@***.***> wrote:

> One thing that's jumped out at me: you can't create a tsdb index with a
> fully empty mapping. You need the @timestamp field declared in the
> mapping to or else we can't stand up the index sorting.
>
> Can you consider using time_series_mode in datastream?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/75638#issuecomment-890652642>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABUXIULC3HTNPV4CN5V5Y3T2X3INANCNFSM5A2M2XQQ>
> .
>

nik9000 committed:
Merge branch 'master' into tsid_gen

nik9000 committed:
mode=time_series

nik9000 commented on server/src/main/java/org/elasticsearch/indices/TimeSeriesIdGeneratorService.java:
```diff
@@ -0,0 +1,386 @@
+        }
+    }
+
+    private abstract static class Value {
```
:+1:

nik9000 committed:
Better names

nik9000 committed:
fixup

nik9000 commented:
@imotov I think this is ready for another round. I believe I've hit all but one of your points - the other one I don't want to do.

I still need to do a datastream test.

nik9000 commented:
Stuff that I'd like to do in a follow up:

* [ ] When backporting this add code to prevent you from disabling soft deletes. We're just not going to have good test coverage for that.
* [ ] Add a field to `IndexRequest` for the timeSeriesId
* [ ] Fix or intentionally disable shard splitting
* [ ] Don't leak the encoded reference into the scripting
* [ ] Figure out what to do about `_id` - should we disable folks using it?
* [ ] Reduce code duplication around routing in Metadata

nik9000 commented:
We should decide two things:

Should indexing to an index in `mode:time_series` without a @timestamp fail? Should an alias with search_routing fail?

nik9000 committed:
Merge branch 'master' into tsid_gen

nik9000 committed:
Cleanup

nik9000 committed:
data stream test

nik9000 committed:
snapshot/restore

imotov commented:
> Should indexing to an index in mode:time_series without a @timestamp fail? 

Good question. Here is a thought - if we make this timestamp mandatory for the mapping we can just add it to the mapping automatically if it is not present in the mapping for all indices with mode:time_series. Then if we cannot find timestamp in a record, we can populate it automatically based on the current time on the node. If we will also treat no dimension as a special null dimension we can theoretically enable mode:time_series on pretty much any index without any modification as long as it doesn't have index sorting or routing required, which might make testing of it a bliss (we just randomly sprinkle it in all our test in create index if we don't have routing and index sorting).

> Should an alias with search_routing fail?

Yes, I think so.

nik9000 committed:
Merge branch 'master' into tsid_gen

nik9000 commented:
> Good question. Here is a thought - if we make this timestamp mandatory for the mapping we can just add it to the mapping automatically if it is not present in the mapping for all indices with mode:time_series. Then if we cannot find timestamp in a record, we can populate it automatically based on the current time on the node. If we will also treat no dimension as a special null dimension we can theoretically enable mode:time_series on pretty much any index without any modification as long as it doesn't have index sorting or routing required, which might make testing of it a bliss (we just randomly sprinkle it in all our test in create index if we don't have routing and index sorting).

Assuming an empty timestamp means `now` feels a little too magic to me. It wouldn't be in the `_source` but it'd be indexed and in doc values like an index time scripted field, but one you didn't create. I'm scared of it, a little.

The null dimension thing feels a little dangerous too - if you don't make any dimensions you are probably making a mistake in how you index. We'll dump all of the `null` dimensions onto one shard. So instead of getting an error message back when you leave out the dimensions you get a very hot shard....

> > Should an alias with search_routing fail?
> 
> Yes, I think so.

Will do.

nik9000 committed:
Alias tests

imotov commented:
> . I'm scared of it, a little.

Yep. It is a bit scary. I guess we have an answer then - we should reject it.

nik9000 committed:
Merge branch 'master' into tsid_gen

nik9000 committed:
Merge branch 'master' into tsid_gen

nik9000 committed:
Remove todo

nik9000 committed:
Moar test

weizijun commented:
Hi, @nik9000, I found a bad case in dynamic template, I push a request to you, https://github.com/nik9000/elasticsearch/pull/9.

when a filed with dimendsion=true in dynamic template, but not in properties, _tsid will not use this field.
but when the second time, the new field added in properties, _tsid will use this field.
It will case the different routing and _tsid.

nik9000 commented:
> Hi, @nik9000, I found a bad case in dynamic template, I push a request to you, [nik9000#9](https://github.com/nik9000/elasticsearch/pull/9).
> 
> when a filed with dimendsion=true in dynamic template, but not in properties, _tsid will not use this field.
> but when the second time, the new field added in properties, _tsid will use this field.
> It will case the different routing and _tsid.

I replied on your PR, but I'll reply here too. Thank! I'm going to convert this into a yaml test case so we automatically run it in a bunch of extra contexts and have a look.

I'm not sure what's up with it yet.

nik9000 committed:
Merge branch 'master' into tsid_gen

nik9000 committed:
WIP

nik9000 committed:
Merge branch 'master' into tsid_gen

nik9000 committed:
WIP

nik9000 committed:
WIP

imotov commented on server/src/main/java/org/elasticsearch/index/mapper/RoutingFieldMapper.java:
```diff
@@ -102,6 +103,11 @@ public boolean required() {
     public void preParse(DocumentParserContext context) {
         String routing = context.sourceToParse().routing();
+        if (context.indexSettings().mode() == IndexMode.TIME_SERIES) {
+            // TODO when we stop storing the tsid in the routing fail any request with routing in time series mode
```
I have been thinking about that. Allowing routing in the time_series mode might actually make sense. Yes, it is allowing users to shoot themselves in the foot. but routing is an advanced feature that already enables screwing  things up on "standard" indices, but it also enables advanced users to build solutions they wouldn't be able to build otherwise.

Having only few elements of TSID in the routing key doesn't break anything, but it can speed things up and help in case of multi-tenant solutions.

nik9000 replied:
I suppose so. I can revive support for routing once we have a field on `IndexRequest` specifically for the tsid. That was something I was hoping to do in a follow up.

imotov commented on server/src/main/java/org/elasticsearch/index/IndexSortConfig.java:
```diff
@@ -112,12 +112,31 @@ private static MultiValueMode parseMultiValueMode(String value) {
+            }
+            if (INDEX_SORT_MISSING_SETTING.exists(settings)) {
+                throw new IllegalArgumentException("Can't set [" + INDEX_SORT_MISSING_SETTING.getKey() + "] in time series mode");
+            }
```
Feels like it can be replaced with something like 
```
            for (Setting<?> setting: indexMode.unsupportedSettings()) {
                if (setting.exists(settings)) {
                    throw new IllegalArgumentException("Can't set [" + setting.getKey() + "] in time series mode");
                }
            }
```

or moved completely into `mode.checkSettings(settings)`

nik9000 replied:
:+1:

imotov reviewed (commented):
I did another pass and left a couple of minor comments. Besides disabling specifying dimensions in dynamic mappings that we discussed and possibly dealing with `copy_to` into dimension fields, I think it looks good. It would be great if somebody from @elastic/es-distributed  could take a look at it as well.

nik9000 committed:
WIP

nik9000 committed:
Merge branch 'master' into tsid_gen

nik9000 committed:
equals/hashCode

nik9000 committed:
Fail nicely

nik9000 committed:
Fix settings

nik9000 commented:
> I am now sure what's up. Its..... icky. The dynamic mapping runs on the data node and doesn't tell the coordinating node what to do. I think the plan for _this_ PR is accept that its broken and then try and fix it in a follow up. Alternatively, we could disallow dynamic mapping adding dimensions. That might be ok in the short term, but I think folks _need_ some way of dynamically adding dimensions and dynamic mappings are the most sensible way to do it in the framework we have now. So we'd have to get it working sooner or later.

I've pushed a change to detect when we modify the list of dimensions with a dynamic mapping update and reject the document. That'll have to do in the short term. In the long term we'll have to make it work but that's a bigger thing. Big enough that we talked about taking another approach to this change entirely. But, we believe the the majority of the change required to get dynamic mappings working lines up with the change required to get hard time bounds on indices working - so we're going to do it anyway. Probably.

My plan is to scrounge up more reviews for this ASAP and then get this in and start on the dynamic mapping updates.

nik9000 committed:
Better

nik9000 committed:
Update tests

nik9000 committed:
javadocs

nik9000 committed:
Revert noops

nik9000 committed:
cleanup

nik9000 committed:
Merge branch 'master' into tsid_gen

nik9000 committed:
More docs

nik9000 committed:
Merge branch 'master' into tsid_gen

nik9000 committed:
Merge branch 'master' into tsid_gen

nik9000 committed:
Merge branch 'master' into tsid_gen

nik9000 committed:
Merge branch 'master' into tsid_gen

nik9000 commented:
For anyone following along from home - it took me a while to get good reviews on this, but we found enough "fun" corner cases involving races with mapping that we're rethinking this. We're unlikely to merge this code as it stands, but fairly likely to use it as an inspiration that will be pretty similar in practice. Just without the races.
